integrations/clickup/clickup/client.py

```diff
@@ -0,0 +1,172 @@
+            f"[ID: {webhook_create['id']}, Team ID: {team_id}]"
+        )
+
+    async def get_paginated_teams(
+        self, params: dict[str, Any] = {}
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        teams_response = await self.client.get(
+            f"{self.clickup_url}/team", params=params
+        )
+        teams_response.raise_for_status()
+        teams = teams_response.json()["teams"]
+        yield teams
+
+    async def _get_paginated_spaces(
+        self, team_id: str, params: dict[str, Any] = {}
+    ) -> list[dict[str, Any]]:
+        spaces_response = await self.client.get(
+            f"{self.clickup_url}/team/{team_id}/space", params=params
+        )
+        spaces_response.raise_for_status()
+        return spaces_response.json()["spaces"]
+
+    async def get_paginated_projects(
+        self, params: dict[str, Any] = {}
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        # getting all teams so as to retrieve the spaces within each team, then using their
+        # ids to get the projects (lists) within each space
+        teams = self.get_paginated_teams(params)
+        async for team_batch in teams:
+            for team in team_batch:
+                spaces = await self._get_paginated_spaces(team["id"], params)
+                projects: list[dict[str, Any]] = []
+
+                for space in spaces:
+                    projects_response = await self.client.get(
+                        f"{self.clickup_url}/space/{space['id']}/list", params=params
+                    )
+                    projects_response.raise_for_status()
+                    # because the port-app-config uses the team ID to relate the projects to the teams
+                    # we add the team ID to each project
+                    # also, the clickup returns the datetime as a timestamp, so we convert it to a
+                    # datetime object
+                    projects.extend(
+                        map(
+                            lambda project: {
+                                **project,
+                                "team_id": team["id"],
+                                "start_date": timestamp_to_datetime(
+                                    project["start_date"]
+                                ),
+                                "due_date": timestamp_to_datetime(project["due_date"]),
+                            },
+                            projects_response.json()["lists"],
+                        )
+                    )
+                yield projects
+
+    async def get_single_project(self, project_id: str) -> dict[str, Any]:
+        # Clickup does not provide a direct way to get the team ID from a project (list), so instead
+        # of directly getting the project, we get all the projects and filter by the project ID
+        # using the result of the get_paginated_projects method which includes the team ID
+        projects = self.get_paginated_projects()
```

Comment:
Since we're calling self.get_paginated_projects() more than once, we need to add the @cache_iterator_result() decorator on it to cache the response